### PR TITLE
R refactor

### PR DESCRIPTION
**Pull request by PavelBal**

### CODE REVIEW

**Comment by PavelBal** [comment]
Fixes https://github.com/Open-Systems-Pharmacology/rClr/issues/28, Fixes https://github.com/Open-Systems-Pharmacology/rClr/issues/27

File: R/messages.R [hunk 1]
```diff
@@ -0,0 +1,10 @@
+messages <- list()
```
Comment (PavelBal) [inline R/messages.R:1]: Following the same concept we have in osp packages, message strings should be in a separate file.

File: R/rClr-exported.R [hunk 2]
```diff
@@ -5,27 +5,7 @@
...
-#' \dontrun{
-#' library(rClr)
-#' ## R.NET is currently used to convert complicated CLR types to sensible R equivalents
-#' setRDotNet()
-#' cTypename <- "Rclr.TestCases"
-#' clrCallStatic(cTypename, "CreateStringDictionary")
-#' setRDotNet(FALSE)
-#' clrCallStatic(cTypename, "CreateStringDictionary")
-#' }
-setRDotNet <- function(setit = TRUE) {
```
Comment (PavelBal) [inline R/rClr-exported.R:27]: Seems to always be on. Moved turing on to package initialization, I do not see a reason why this function should be accessable for user.

File: R/rClr-exported.R [hunk 3]
```diff
@@ -71,7 +51,7 @@ setConvertAdvancedTypes <- function(enable = TRUE) {
 #' }
 clrLoadAssembly <- function(name) {
   # if( !file.exists(name) ) stop(paste("File not found: ", name))
-  result <- .C("rclr_load_assembly", name, PACKAGE = nativePkgName)
+  result <- .C("rclr_load_assembly", name, PACKAGE = rNETenv$nativePkgName)
```
Comment (PavelBal) [inline R/rClr-exported.R:54]: Package name is stored in the environment and not as a global variable.

File: R/rClr-exported.R [hunk 4]
```diff
@@ -687,9 +667,6 @@ inspectArgs <- function(...) {
 #' @return A string
 #' @export
 clrVT <- function(objOrType, methodName, ...) {
-  if (nativePkgName != "rClrMs") {
```
Comment (PavelBal) [inline R/rClr-exported.R:690]: We should not have any cases where this is not supported once moved to .NET7.

File: R/rClr-internal.R [hunk 5]
```diff
@@ -4,33 +4,10 @@ reflectionHelperTypeName <- "Rclr.ReflectionHelper"
...
-#'
-#' Initialize a new CLR application domain
-#'
-#' @param debug If using Mono, should the CLR be initialised to try to hook up to the mono soft debugger in MonoDevelop. This parameter has no effect if using MS.NET.
-#' @return nothing is returned by this function
-#' @examples
-#' \dontrun{
-#' library(rClr)
-#' }
-clrInit <- function(debug = FALSE) {
```
Comment (PavelBal) [inline R/rClr-internal.R:17]: Moved directly to zzz.R, not required as a separate function.

File: R/rClr-internal.R [hunk 6]
```diff
@@ -4,33 +4,10 @@ reflectionHelperTypeName <- "Rclr.ReflectionHelper"
...
-  f <- path.expand(f)
-  stopifnot(file.exists(f))
-  result <- .C("rclr_create_domain", pkgLibsDir, f, as.integer(debug), PACKAGE = nativePkgName)
-}
-
 checkIsExtPtr <- function(extPtr) {
   stopifnot("externalptr" %in% class(extPtr))
 }
 
-getLibsPath <- function(pkgName) {
```
Comment (PavelBal) [inline R/rClr-internal.R:29]: Not required any more.

File: R/rNET-env.R [hunk 7]
```diff
@@ -0,0 +1,36 @@
+# Environment that holds various global variables and settings for the package.
+# It is not exported and should not be directly manipulated by other packages.
+rNETenv <- new.env(parent = emptyenv())
```
Comment (PavelBal) [inline R/rNET-env.R:3]: Created an environment following the same concept we have in other osp packages.

File: R/zzz.R [hunk 8]
```diff
@@ -1,112 +1,45 @@
-# An internal variable that is set ot the name of the native library depending on its use of the Mono or MS.NET CLR
-nativePkgName <- ""
```
Comment (PavelBal) [inline R/zzz.R:2]: This is stored in the env, no evil global variables any more.

File: R/zzz.R [hunk 9]
```diff
@@ -1,112 +1,45 @@
...
-
-# An internal variable to buffer startup messages
-startupMsg <- ""
```
Comment (PavelBal) [inline R/zzz.R:5]: This was the message on startup: 

> Loading the dynamic library for Microsoft .NET runtime...
Loaded Common Language Runtime version 4.0.30319.42000

I don't think we should by default print this info on load. Runtime version can still be accessed from the env.

File: R/zzz.R [hunk 10]
```diff
@@ -1,112 +1,45 @@
...
-          sep = "\n"
-        ))
-      }
-      if (!(msDll %in% dlls)) {
-        stop(paste("rClr library .NET framework not found - looked under", archLibPath, "but not found in", paste(dlls, collapse = ",")))
-      }
-      appendStartupMsg("Loading the dynamic library for Microsoft .NET runtime...")
-      chname <- getFnameNoExt(msDll)
-      loadAndInit(chname, pkgname, libname, srcPkgLibPath)
+    if (Sys.which(rNETenv$msvcrFileName) == "") {
```
Comment (PavelBal) [inline R/zzz.R:5]: Should we add a similar check for linux? Well once migrated to .NET7

**Comment by PavelBal** [comment]
Went through the loading section.

File: R/rNET-env.R [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+
+# name of the package. This will be used to retrieve information on the package at run time
+rNETenv$packageName <- "rClr"
+# Name of the native library
+rNETenv$nativePkgName <- "rClrMs"
```
Comment (Yuri05) [inline R/rNET-env.R:8]: package names to be discussed. But let's keep it AS IS for this PR

Comment (PavelBal) [reply R/rNET-env.R:8]: Exactly.

File: R/zzz.R [hunk 12]
```diff
@@ -1,112 +1,45 @@
...
+      stop(paste(rNETenv$msvcrFileName, "was not found on this Windows system.",
+        "You are probably missing the Visual C++ Redistributable.",
+        "Go to https://learn.microsoft.com/en-US/cpp/windows/latest-supported-vc-redist?view=msvc-170 and dowload the latest 'Microsoft Visual C++ Redistributable'",
```
Comment (Yuri05) [inline R/zzz.R]: typo: `dowload `

Comment (PavelBal) [inline R/zzz.R]:
```suggestion
        "Go to https://learn.microsoft.com/en-US/cpp/windows/latest-supported-vc-redist?view=msvc-170 and download the latest 'Microsoft Visual C++ Redistributable'",
```

File: R/zzz.R (hunk 10 above)
Comment (Yuri05) [reply R/zzz.R:5]: In Linux we should check for the glibc version. E.g.

```
> system("ldd --version | grep ldd", intern = TRUE)
[1] "ldd (Ubuntu GLIBC 2.27-3ubuntu1.6) 2.27"
>
```

Once we know which glibc version is required for the package: the installed version must be greater or equal.
E.g.  in the example above, if we require glibc 2.26 or 2.27 - all good, if we require glibc 2.28 - problem.

File: R/zzz.R [hunk 13]
```diff
@@ -1,112 +1,45 @@
...
 }
 
-appendStartupMsg <- function(msg) {
-  startupMsg <<- paste0(startupMsg, msg, "\n")
+.loadAndInit <- function() {
+  # Path to the folder where the libraries are located
+  srcPkgLibPath <- system.file("libs", package = rNETenv$packageName)
+
+  # Load C++ library
+  f <- file.path(srcPkgLibPath, paste0(rNETenv$nativePkgName, .Platform$dynlib.ext))
```
Comment (Yuri05) [inline R/zzz.R]: under Linux, the `lib `prefix is missing.
`MyLib.dll` ==> `libMyLib.so`

Comment (PavelBal) [reply R/zzz.R]: Is it always `lib<Something>.so`?

Comment (Yuri05) [reply R/zzz.R]: yes

File: R/zzz.R (hunk 12 above)
Comment (PavelBal) [reply R/zzz.R]: Fixed.

File: R/zzz.R (hunk 13 above)
Comment (PavelBal) [reply R/zzz.R]: Will this work?

**Review by Yuri05 — APPROVED** [review]

**Pull request merged**